### PR TITLE
CRM-21671 abstract core task

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
This adds a new CRM_Core_Task class which all the other Task classes can inherit from as it's mostly duplicated code.
Split from CRM-21391 so it can be reviewed more easily.

Before
----------------------------------------
Lot's of duplicated code

After
----------------------------------------
Once CRM-21391 is merged, mostly shared code - easier to maintain and less buggy :-)

---

 * [CRM-21671: Add core tasks class](https://issues.civicrm.org/jira/browse/CRM-21671)
 * [CRM-21391: Refactor tasks to use a base class](https://issues.civicrm.org/jira/browse/CRM-21391)

### CODE REVIEW

**Comment by mattwire** [comment]
Jenkins test this please

**Comment by seamuslee001** [comment]
@mattwire error is 

```
PHP Parse error: syntax error, unexpected 'PRINT' (T_PRINT), expecting identifier (T_STRING) in CRM/Core/Task.php on line 44 Errors parsing CRM/Core/Task.php```

**Comment by eileenmcnaughton** [comment]
I merged the other one so might need a clean up - can you include just one of the child classes in here too ?

**Comment by mattwire** [comment]
Jenkins test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by mattwire** [comment]
@eileenmcnaughton Yay, passes tests :-)

File: CRM/Core/Task.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  /**
+   * The task array
+   *
+   * @var array
+   */
+  public static $_tasks = NULL;
+
+  static $objectType = NULL;
+
+  public static function tasks() {
```
Comment (colemanw) [inline CRM/Core/Task.php]: This function could use a docblock.

Comment (mattwire) [reply CRM/Core/Task.php]: @colemanw Agreed.  I've gone through and added more comments/docblocks

**Comment by eileenmcnaughton** [comment]
I read through this code & tested doing a participant search & choosing various actions. I hit no issues and the code is clearly an improvement.

I think we are OK without a unit test as this is very form layer and there is no clear path to add tests here - although it might be nice to think of some

**Comment by kngs** [comment]
After a recent upgrade from 4.7.29 to 5.2.1 (Joomla) I experienced the following problem:

After doing an Events/Find Participants search and selecting some or all participants, the "Print selected rows" option in no longer appearing in the Actions dropdown.

An Advanced Search when displaying results as Event Participants also is missing this option. The option does appear when results are displayed as Contacts.

I have verified the same behavior on dmaster.demo (running 5.4.alpha1).

Issue filed by Pradeep: https://lab.civicrm.org/dev/core/issues/185

**Pull request merged**